save-plugins/fix-and-warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugins/fixandwarn/FixAndWarnPlugin.kt

```diff
@@ -0,0 +1,181 @@
+    useInternalRedirections) {
+    private val fixPluginConfig = testConfig.pluginConfigs.filterIsInstance<FixAndWarnPluginConfig>().single().fixPluginConfig
+    private val warnPluginConfig = testConfig.pluginConfigs.filterIsInstance<FixAndWarnPluginConfig>().single().warnPluginConfig
+    private val fixPlugin = FixPlugin(
+        createTestConfigForPlugins(TestConfigSections.FIX),
+        testFiles
+    )
+    private val warnPlugin = WarnPlugin(
+        createTestConfigForPlugins(TestConfigSections.WARN),
+        testFiles
+    )
+
+    /**
+     * Create TestConfig same as current, but with corresponding plugin configs list for nested [fix] and [warn] sections
+     *
+     * @param type type of nested section
+     * @return TestConfig for corresponding section
+     */
+    private fun createTestConfigForPlugins(type: TestConfigSections) = TestConfig(
+        testConfig.location,
+        testConfig.parentConfig,
+        mutableListOf(
+            testConfig.pluginConfigs.filterIsInstance<GeneralConfig>().single(),
+            if (type == TestConfigSections.FIX) {
+                fixPluginConfig
+            } else {
+                warnPluginConfig
+            }
+        )
+    )
+
+    override fun handleFiles(files: Sequence<List<Path>>): Sequence<TestResult> {
+        testConfig.validateAndSetDefaults()
+
+        // Test files for warn plugin could be obtained by filtering of `fix` resources: excluding `expected` files
+        val testFilePattern = warnPluginConfig.resourceNamePattern
+        val warnTestFiles = files.filterTestResources(testFilePattern, match = true)
+
+        // Warn plugin should process the files, which were fixed by fix plugin
+        // We can get the paths of fixed files from warn test files (since they are the same for fix plugin),
+        // and then just add relative path from FixPlugin tmp dir
+        val testFilesAfterFix: MutableList<List<Path>> = mutableListOf()
+        warnTestFiles.forEach { path ->
+            // TODO change location from hardcoded FixPlugin::simpleName after https://github.com/cqfn/save/issues/156
+            testFilesAfterFix.add(listOf(constructPathForCopyOfTestFile(FixPlugin::class.simpleName!!, path)))
+        }
+
+        logDebug("FixPlugin test resources: ${files.toList()}")
+        logDebug("WarnPlugin test resources: $testFilesAfterFix")
+
+        val expectedFiles = files.filterTestResources(testFilePattern, match = false)
+
+        // Remove (in place) warnings from test files before fix plugin execution
+        val filesAndTheirWarningsMap = removeWarningsFromExpectedFiles(expectedFiles)
+
+        val fixTestResults = fixPlugin.handleFiles(files)
+
+        // Fill back original data with warnings
+        filesAndTheirWarningsMap.forEach { (filePath, warningsList) ->
+            val fileData = fs.readLines(filePath) as MutableList
+            warningsList.forEach { (line, warningMsg) ->
+                fileData.add(line, warningMsg)
```

Comment:
So ,you simply append comments with warnings to the end of the file? What if comments have dynamically calculated line number, like we will support in #159 and #160 ?
We have a related issue #133. and logic can be implemented for both Fix and FixAndWarn plugins there. Maybe, we should fix file *with* comments and then remove them when *comparing* with expected? This way we can directly check fixed file with warn-plugin, and if they use placeholders for line number, they don't need to keep in mind line changes.

Reply:
Actually, not really. `warningsList` contain the pairs of line number of warning and warning message from the original version of expected file.
So, since in `fileData` there is a content of expected file without any warnings, and also we know (by warningsList) on which lines there were warnings, we just add them into appropriate place. Here I used the overload of `add` function, which `Inserts an element into the list at the specified index`.

About `What if comments have dynamically calculated line number` - I think it is not connected with current algorithm, since here I save the line not of problematic code on which we should raise warn, but the line of warning itself.

Don't really understand how we can check for expected warnings if user won't provide them to us `users can write expected files without these comments`.

Reply:
Oh, didn't see, that it's `add` with index, sorry.

> Don't really understand how we can check for expected warnings if user won't provide them to us users can write expected files without these comments.

I mean, that only `*Test` files should contain comments with warnings, because we can implement comparison with expected files with these comments removed. This will allow users not to write same comments twice. And warn-plugin needs to operate on mutated `*Test` file anyway.

Reply:
Discussed offline. For now warn plugin will check warnings in `Expected` files